### PR TITLE
Fixed #27083 -- Added basic support for weak ETags.

### PR DESCRIPTION
**Pull request by marfire**
https://code.djangoproject.com/ticket/27083

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by charettes** [comment]
The tests are failing on Python 2.7 because of your usage of `fullmatch`.

**Comment by marfire** [comment]
@charettes: Thanks, I've updated the PR and started testing with 2.7 locally.

I also added additional tests and documentation. Still missing is an addition to the release notes.

**Comment by marfire** [comment]
By the way, apologies if I've been spamming the build server. I was using my Github fork to pass work between two servers (dev and test) without realizing that each push updated the PR.

**Comment by timgraham** [comment]
If it's possible to break this up into some smaller commits, that could make review much easier, especially for a non-ETag expert like me. :-)

**Comment by marfire** [comment]
Ha, this *was* my attempt to keep things small! :-O (As I'm also planning to address the ETag issues discussed [here](https://code.djangoproject.com/ticket/19705#comment:14)...)

I think this could still reasonably be divided into two parts: one adding support for the use of weak ETags, and the other updating the `get_conditional_response()` algorithm to match RFC 7232. Does that sound sensible?

The reason I did both at the same time is that adding weak ETag support required changing `get_conditional_response()` anyway, so I thought I'd go ahead and change it fully rather than update the old version just to throw it away with the next commit. But I don't mind splitting it if you'd find the review easier...

**Comment by timgraham** [comment]
Whatever you think, I don't want to waste your time.

**Comment by marfire** [comment]
OK, I split the commit. This one adds syntactic support for weak ETags but doesn't change the complicated comparison algorithm.

It touches a fair number of lines, but many of the changes are trivial (e.g. removing quotes from comparisons in the tests). Hopefully it will be easier to review...

**Comment by timgraham** [comment]
Is it ready for review then? (Not sure since the ticket flags aren't updated.)

**Comment by marfire** [comment]
Good point, I neglected to do that. It's ready, and I will update the flags.

As I mentioned above, I'm not sure what merits a mention in the release notes. There is technically a backwards incompatibility here (see the last paragraph of the ticket description). Should I put a mention under Backwards incompatible changes in 1.11 -> Miscellaneous?

**Comment by timgraham** [comment]
Sure, unless there's more a sentence of two of explanation in which case you could create your own heading.

File: django/utils/http.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
     urlparse,
 )
 
-ETAG_MATCH = re.compile(r'(?:W/)?"((?:\\.|[^"])*)"')
+ETAG_MATCH = re.compile(r'\A((?:W/)?"[^"]*")\Z')  # based on RFC 7232, Appendix C
```
Comment (timgraham) [inline django/utils/http.py]: Could you annotate the regex similar to:
```
tld_re = (
    '\.'                                # dot
    '(?!-)'                             # can't start with a dash
    '(?:[a-z' + ul + '-]{2,63}'         # domain label
    '|xn--[a-z0-9]{1,59})'              # or punycode label
    '(?<!-)'                            # can't end with a dash
    '\.?'                               # may have a trailing dot
)
```

File: django/utils/http.py [hunk 2]
```diff
@@ -90,7 +90,7 @@ def urlencode(query, doseq=0):
         query = query.items()
     return original_urlencode(
         [(force_str(k),
-         [force_str(i) for i in v] if isinstance(v, (list, tuple)) else force_str(v))
+          [force_str(i) for i in v] if isinstance(v, (list, tuple)) else force_str(v))
```
Comment (timgraham) [inline django/utils/http.py]: It would be better not to make whitespace changes like this.

File: django/utils/http.py [hunk 3]
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
 
 def parse_etags(etag_str):
     """
-    Parses a string with one or several etags passed in If-None-Match and
-    If-Match headers by the rules in RFC 2616. Returns a list of etags
-    without surrounding double quotes (") and unescaped from \<CHAR>.
+    Parses a string of ETags given in an If-None-Match or If-Match header as
```
Comment (timgraham) [inline django/utils/http.py]: When changing docstrings, we are updating them to use PEP 257 verb style: Parse / Return, etc.

File: django/utils/http.py [hunk 4]
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
...
-        return [etag_str]
-    etags = [e.encode('ascii').decode('unicode_escape') for e in etags]
-    return etags
+    if etag_str.strip() == '*':
+        return ['*']
+    else:
+        # This is just enough validation to prevent outdated RFC 2616 ETags
+        # from breaking things.
+        etag_matches = (ETAG_MATCH.match(etag.strip()) for etag in etag_str.split(","))
+        etags = [match.group(1) for match in etag_matches if match]
```
Comment (timgraham) [inline django/utils/http.py]: return without intermediate `etags` variable

File: django/utils/http.py [hunk 5]
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
...
-    etags = ETAG_MATCH.findall(etag_str)
-    if not etags:
-        # etag_str has wrong format, treat it as an opaque string then
-        return [etag_str]
-    etags = [e.encode('ascii').decode('unicode_escape') for e in etags]
-    return etags
+    if etag_str.strip() == '*':
+        return ['*']
+    else:
+        # This is just enough validation to prevent outdated RFC 2616 ETags
```
Comment (timgraham) [inline django/utils/http.py]: prevent ETags in the old RFC 2616 format

File: django/utils/http.py [hunk 6]
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
...
+        # from breaking things.
```
Comment (timgraham) [inline django/utils/http.py]: Perhaps "breaking things" could be better described. Is the idea that they should still be matched?

File: django/utils/http.py [hunk 7]
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
...
+        etag_matches = (ETAG_MATCH.match(etag.strip()) for etag in etag_str.split(","))
```
Comment (timgraham) [inline django/utils/http.py]: use single quotes for consistency

File: django/utils/http.py [hunk 8]
```diff
@@ -232,30 +232,31 @@ def urlsafe_base64_decode(s):
...
     """
-    Unquote an ETag string; i.e. revert quote_etag().
+    If the provided string is already a quoted ETag, return it. Otherwise, wrap
+    the string in quotes, making it a strong ETag.
     """
-    return etag.strip('"').replace('\\"', '"').replace('\\\\', '\\') if etag else etag
+    if ETAG_MATCH.match(etag_str):
+        return etag_str
+    else:
+        # We don't validate etag_str.
```
Comment (timgraham) [inline django/utils/http.py]: I'm trying to avoid "We" in comments. I think "Don't validate etag_str." has the same meaning.

File: django/views/decorators/http.py [hunk 9]
```diff
@@ -58,20 +58,20 @@ def condition(etag_func=None, last_modified_func=None):
 
     The parameters are callables to compute the ETag and last modified time for
     the requested resource, respectively. The callables are passed the same
-    parameters as the view itself. The Etag function should return a string (or
+    parameters as the view itself. The ETag function should return a string (or
```
Comment (timgraham) [inline django/views/decorators/http.py]: It would be appropriate to make the renames in a separate commit unless this is the only one to be fixed in which case it's probably not worth separating it.

File: django/middleware/http.py [hunk 10]
```diff
@@ -11,6 +11,7 @@ class ConditionalGetMiddleware(MiddlewareMixin):
 
     Also sets the Date and Content-Length response-headers.
     """
+
```
Comment (timgraham) [inline django/middleware/http.py]: Please revert whitespace change.

File: docs/releases/1.11.txt [hunk 11]
```diff
@@ -483,6 +483,10 @@ Miscellaneous
   ``migrate`` command. Use the new :djadmin:`remove_stale_contenttypes` command
   instead.
 
+* ETags are now parsed according to the updated RFC 7232 Conditional Requests
```
Comment (timgraham) [inline docs/releases/1.11.txt]: RFC 7232 ->
```
:rfc:`7232`
```

File: docs/releases/1.11.txt [hunk 12]
```diff
@@ -483,6 +483,10 @@ Miscellaneous
...
+  specification. The ETag syntax there is slightly different from that of the
+  older RFC 2616 specification which was used in previous versions.
```
Comment (timgraham) [inline docs/releases/1.11.txt]: Is it worth giving an example of how this might affect a project or will it be obvious to ETag users?

File: docs/topics/conditional-view-processing.txt [hunk 13]
```diff
@@ -64,7 +64,15 @@ order, as the view function they are helping to wrap. The function passed
 ``last_modified_func`` should return a standard datetime value specifying the
 last time the resource was modified, or ``None`` if the resource doesn't
 exist. The function passed to the ``etag`` decorator should return a string
-representing the `Etag`_ for the resource, or ``None`` if it doesn't exist.
+representing the `ETag`_ for the resource, or ``None`` if it doesn't exist.
+
+.. versionchanged:: 1.11
+
+    In prior versions the return value from ``etag_func()`` was interpreted as
```
Comment (timgraham) [inline docs/topics/conditional-view-processing.txt]: In older versions,

File: docs/topics/conditional-view-processing.txt [hunk 14]
```diff
@@ -64,7 +64,15 @@ order, as the view function they are helping to wrap. The function passed
...
+    the unquoted part of the ETag. That prevented the use of weak ETags, which
+    have the format ``W/"<string>"``. The return value is now expected to be
+    an ETag as defined by the specification (including the quotes), although
+    the unquoted format is also accepted for backwards compatibility reasons.
```
Comment (timgraham) [inline docs/topics/conditional-view-processing.txt]: Is it fine to remove this entire note when removing or should anything remain? If the latter, please make the note self-contained as described in https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features.

File: tests/conditional_processing/tests.py [hunk 15]
```diff
@@ -24,12 +25,16 @@ def assertFullResponse(self, response, check_last_modified=True, check_etag=True
...
             self.assertEqual(response['Last-Modified'], LAST_MODIFIED_STR)
         if check_etag:
-            self.assertEqual(response['ETag'], '"%s"' % ETAG)
+            self.assertEqual(response['ETag'], ETAG)
 
     def assertNotModified(self, response):
         self.assertEqual(response.status_code, 304)
         self.assertEqual(response.content, b'')
 
+    def assertPreconditionFailed(self, response):
```
Comment (timgraham) [inline tests/conditional_processing/tests.py]: Perhaps this helper could be added in a separate commit so there are less unrelated test changes in this commit (unless it really is related).

File: tests/conditional_processing/tests.py [hunk 16]
```diff
@@ -188,7 +193,32 @@ def test_single_condition_head(self):
         response = self.client.head('/condition/')
         self.assertNotModified(response)
 
+    # Test that the same ETag results whether returning quoted
```
Comment (timgraham) [inline tests/conditional_processing/tests.py]: Should these comments be docstrings? We're in the habit of stating the expected behavior rather than including "Test that..." since all tests test things.

File: tests/utils_tests/test_http.py [hunk 17]
```diff
@@ -196,14 +196,22 @@ def test_is_same_domain_bad(self):
...
 class ETagProcessingTests(unittest.TestCase):
     def test_parsing(self):
-        etags = http.parse_etags(r'"", "etag", "e\"t\"ag", "e\\tag", W/"weak"')
-        self.assertEqual(etags, ['', 'etag', 'e"t"ag', r'e\tag', 'weak'])
+        etags = http.parse_etags(r'"" ,  "etag", "e\\tag", W/"weak"')
+        self.assertEqual(etags, ['""', '"etag"', r'"e\\tag"', 'W/"weak"'])
+        etags = http.parse_etags('*')
+        self.assertEqual(etags, ['*'])
+
+        # We ignore RFC 2616 tags that are now invalid.
```
Comment (timgraham) [inline tests/utils_tests/test_http.py]: Chop "We"
"now invalid" ->  "invalid according to RFC 7232"

File: tests/conditional_processing/views.py [hunk 18]
```diff
@@ -27,3 +27,23 @@ def etag_view1(request):
 def etag_view2(request):
     return HttpResponse(FULL_RESPONSE)
 etag_view2 = etag(lambda r: ETAG)(etag_view2)
+
+
+# etag_func() returns an ETag without quotes
```
Comment (timgraham) [inline tests/conditional_processing/views.py]: docstrings?

File: django/utils/http.py (hunk 6 above)
Comment (marfire) [inline django/utils/http.py]: I'll reword this. What I'm getting at is that splitting on `,` allows each tag to be matched or not individually, whereas doing a findall() like before allows one weird RFC 2616 tag to break all the others (e.g. `"\"", "X"` would become `"\"` and `", "`).

File: docs/releases/1.11.txt (hunk 12 above)
Comment (marfire) [inline docs/releases/1.11.txt]: In my opinion an example would be overkill. My guess is that this change won't affect anyone, but if it does it will be because they're supplying their own, unusually-formatted ETags, and in that case they should really be consulting the RFCs directly. Let me know if you disagree.

File: django/middleware/http.py (hunk 10 above)
Comment (timgraham) [inline django/middleware/http.py]: The guideline is "don't include unrelated cosmetic changes in bug/feature patch."

File: docs/topics/conditional-view-processing.txt (hunk 14 above)
Comment (marfire) [inline docs/topics/conditional-view-processing.txt]: To me, the existing phrase "a string representing the ETag" implies the full quoted form as defined by the specification (including weak ETags). So going forward, I think supporting the unquoted version can be an undocumented nod to backwards compatibility, making it fine for this note to go away. Another option would be to spell out the quoted / unquoted thing explicitly in the documentation. Yet another option would be to actively deprecate support for the unquoted form (though it doesn't seem worth the trouble to me).

Let me know your opinion.

Comment (timgraham) [inline docs/topics/conditional-view-processing.txt]: I trust your judgment.

**Comment by marfire** [comment]
Thanks for the review, Tim, I've updated the PR.

File: tests/conditional_processing/tests.py [hunk 19]
```diff
@@ -188,7 +189,34 @@ def test_single_condition_head(self):
...
+    def test_unquoted(self):
+        """
+        The same quoted ETag should be set on the header regardless of whether
+        etag_func() in condition() returns a quoted or an unquoted ETag.
+        """
+        response_quoted = self.client.get('/condition/etag/')
+        response_unquoted = self.client.get('/condition/unquoted_etag/')
+        self.assertEqual(response_quoted['ETag'], response_unquoted['ETag'])
+
+    # Since it's possible that the matching algorithm could use the wrong value
```
Comment (timgraham) [inline tests/conditional_processing/tests.py]: Docstring?

File: tests/conditional_processing/urls.py [hunk 20]
```diff
@@ -8,4 +8,7 @@
     url('^condition/last_modified2/$', views.last_modified_view2),
     url('^condition/etag/$', views.etag_view1),
     url('^condition/etag2/$', views.etag_view2),
+    url('^condition/unquoted_etag/$', views.etag_view_unquoted),
+    url('^condition/weak_etag/$', views.etag_view_weak),
```
Comment (timgraham) [inline tests/conditional_processing/urls.py]: URL seems to be unused

File: docs/topics/conditional-view-processing.txt [hunk 21]
```diff
@@ -64,7 +64,15 @@ order, as the view function they are helping to wrap. The function passed
 ``last_modified_func`` should return a standard datetime value specifying the
 last time the resource was modified, or ``None`` if the resource doesn't
 exist. The function passed to the ``etag`` decorator should return a string
-representing the `Etag`_ for the resource, or ``None`` if it doesn't exist.
+representing the `ETag`_ for the resource, or ``None`` if it doesn't exist.
+
+.. versionchanged:: 1.11
+
+    In older versions the return value from ``etag_func()`` was interpreted as
```
Comment (timgraham) [inline docs/topics/conditional-view-processing.txt]: versions,

File: docs/topics/conditional-view-processing.txt [hunk 22]
```diff
@@ -64,7 +64,15 @@ order, as the view function they are helping to wrap. The function passed
...
+    the unquoted part of the ETag. That prevented the use of weak ETags, which
+    have the format ``W/"<string>"``. The return value is now expected to be
+    an ETag as defined by the specification (including the quotes), although
+    the unquoted format is also accepted for backwards compatibility reasons.
```
Comment (timgraham) [inline docs/topics/conditional-view-processing.txt]: chop "reasons"

File: docs/releases/1.11.txt [hunk 23]
```diff
@@ -488,6 +488,10 @@ Miscellaneous
   ``migrate`` command. Use the new :djadmin:`remove_stale_contenttypes` command
   instead.
 
+* ETags are now parsed according to the updated :rfc:`7232` Conditional Requests
```
Comment (timgraham) [inline docs/releases/1.11.txt]: Perhaps a bit simpler:
```
* ETags are now parsed according to the :rfc:`7232` Conditional Requests
  specification rather than the syntax from :rfc:`2616`.
```

File: tests/conditional_processing/tests.py (hunk 19 above)
Comment (marfire) [inline tests/conditional_processing/tests.py]: I don't think this makes sense as a docstring. It's not explaining what the function does (or tests), it's more of a meta-comment explaining why this test exists at all (as opposed to having no additional tests and relying on `test_unquoted()`, or duplicating all of the tests for both quoted and unquoted values).

File: tests/conditional_processing/urls.py (hunk 20 above)
Comment (marfire) [inline tests/conditional_processing/urls.py]: You're right, it's leftover from the original commit...

**Comment by marfire** [comment]
I've updated the PR.

By the way, instead of amending the commit, would it be better if I rebased the commit and then made the post-review changes as a separate commit? I was trying to save you the trouble of squashing multiple commits in the end, but maybe it makes review easier...

**Comment by timgraham** [comment]
merged in 4ef0e019b7dd3d2bf93b5c705b3b7df9cdb77561, thanks!

For larger PRs, incremental commits can be helpful, but this is small enough that the choice you made is good.
